core/src/main/java/software/amazon/awssdk/auth/profile/Profile.java

```diff
@@ -0,0 +1,288 @@
+     * module in order to generate the credentials provider. If it's not available, an illegal state exception will be raised.
+     */
+    private AwsCredentialsProvider roleBasedProfileCredentialsProvider() {
+        requireProperties(ProfileProperties.SOURCE_PROFILE);
+
+        Validate.validState(credentialsParent != null,
+                            "The profile '%s' must be configured with a credentials parent in order to use assumed roles.", name);
+
+        AwsCredentialsProvider parentCredentialsProvider = credentialsParent.credentialsProvider()
+                                                                            .orElseThrow(this::noParentCredentialsException);
+
+        return stsCredentialsProviderFactory().create(parentCredentialsProvider, this);
+    }
+
+    /**
+     * Require that the provided properties are configured in this profile.
+     */
+    private void requireProperties(String... requiredProperties) {
+        Arrays.stream(requiredProperties)
+              .forEach(p -> Validate.isTrue(properties.containsKey(p),
+                                            "Profile property '%s' was not configured for '%s'.", p, name));
+    }
+
+    private IllegalStateException noParentCredentialsException() {
+        String error = String.format("The parent profile of '%s' was configured to be '%s', but that parent profile has no "
+                                     + "credentials configured.", name, properties.get(ProfileProperties.SOURCE_PROFILE));
+        return new IllegalStateException(error);
+    }
+
+    /**
+     * Load the factory that can be used to create the STS credentials provider, assuming it is on the classpath.
+     */
+    private ChildProfileCredentialsProviderFactory stsCredentialsProviderFactory() {
+        try {
+            Class<?> stsCredentialsProviderFactory = Class.forName(STS_PROFILE_CREDENTIALS_PROVIDER_FACTORY, true,
+                                                                   Thread.currentThread().getContextClassLoader());
+            return (ChildProfileCredentialsProviderFactory) stsCredentialsProviderFactory.getConstructor().newInstance();
+        } catch (ClassNotFoundException e) {
+            throw new IllegalStateException("To use assumed roles in the '" + name + "' profile, the 'sts' service module must "
+                                            + "be on the class path.");
+        } catch (NoSuchMethodException | InvocationTargetException | InstantiationException | IllegalAccessException e) {
+            throw new IllegalStateException("Failed to create the '" + name + "' profile credentials provider.", e);
+        }
+    }
+
+    @Override
+    public Builder toBuilder() {
+        return builder().name(name)
+                        .properties(properties)
+                        .credentialsParent(credentialsParent);
+    }
+
+    @Override
+    public String toString() {
+        return "Profile(" + name + ", " + properties + ")";
```

Comment:
isn't `properties` just a map - will this print nice-like?

Reply:
https://docs.oracle.com/javase/7/docs/api/java/util/AbstractMap.html#toString()